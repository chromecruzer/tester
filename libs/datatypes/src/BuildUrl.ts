import _ from 'lodash';

export default class BuildUrl {
  constructor(private host: (string | null) = null, private port: (string | null) = null) {

  }

  public build(path: string, paths: string[] = [], params = {}) {
    let result = `${this.prefix()}${path}`;
    let firstDelimiter = true;
    paths.forEach(p => {
      result += `${firstDelimiter ? '/' : ''}${p}`;
      firstDelimiter = true;
    });
    if (_.isEmpty(params)) {
      return result;
    }
    // console.log(`result so far "${result}" last char is ${_.last(result)}`)
    result += _.last(result) === '/' ? '?' : '/?';
    firstDelimiter = false;
    _.forEach(params, (value, key) => {
      if (!_.isArray(value)) {
        result += `${firstDelimiter ? '&' : ''}${key}=${value}`;
        firstDelimiter = true;
        return;
      }
      _.forEach(value, v => {
        result += `${firstDelimiter ? '&' : ''}${key}=${v}`;
        firstDelimiter = true;
      });
      // console.log(`result after param ${key}=${value} ${result}`);
    });
    return result;
  }

  private prefix() {
    switch (true) {
      case this.host !== null && this.port !== null:
        return `${this.host}:${this.port}/`;
      case this.host !== null:
        return `${this.host}/`;
      default:
        return '';
    }
  }
}
