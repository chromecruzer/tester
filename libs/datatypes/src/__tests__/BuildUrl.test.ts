import BuildUrl from "../BuildUrl";

describe('BuildUrl', () => {
  describe('build', () => {
    it('should return the prefix if no parameters', () => {
        const builder = new BuildUrl('localhost', '1400');
        const actual = builder.build('subpath');
        expect(actual).toMatchSnapshot();
    });
    it('should return prefix with only host if no port', () => {
      const builder = new BuildUrl('localhost');
      const actual = builder.build('subpath');
      expect(actual).toMatchSnapshot();
    });
    it('should return /path if no host or port', () => {
      const builder = new BuildUrl();
      const actual = builder.build('subpath');
      expect(actual).toMatchSnapshot();
    });
    it('should add path variables in correct order', () => {
      const builder = new BuildUrl();
      const actual = builder.build('subpath', ['path1', 'twopath','threepathpigs']);
      expect(actual).toMatchSnapshot();
    });
    it('should add search parameters correctly', () => {
      const builder = new BuildUrl();
      const actual = builder.build('subpath', [], {key1:'justme', key2:'#pro'});
      expect(actual).toMatchSnapshot();
    });
    it('should add search parameters with array of values correctly', () => {
      const builder = new BuildUrl();
      const actual = builder.build('subpath', [], {array1:['just','me','and','my','boo'], key2:'#pro', array:['tag', 'youre', 'it']});
      expect(actual).toMatchSnapshot();
    });
    it('should adds all parameters correctly', () => {
      const builder = new BuildUrl();
      const actual = builder.build('subpath', ['path1'],{key1:'justme', key2:'#pro'});
      expect(actual).toMatchSnapshot();
    });
  });
});
