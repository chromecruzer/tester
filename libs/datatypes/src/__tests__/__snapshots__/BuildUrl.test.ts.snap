// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildUrl build should add path variables in correct order 1`] = `"subpath/path1/twopath/threepathpigs"`;

exports[`BuildUrl build should add search parameters correctly 1`] = `"subpath/?key1=justme&key2=#pro"`;

exports[`BuildUrl build should add search parameters with array of values correctly 1`] = `"subpath/?array1=just&array1=me&array1=and&array1=my&array1=boo&key2=#pro&array=tag&array=youre&array=it"`;

exports[`BuildUrl build should adds all parameters correctly 1`] = `"subpath/path1/?key1=justme&key2=#pro"`;

exports[`BuildUrl build should return /path if no host or port 1`] = `"subpath"`;

exports[`BuildUrl build should return prefix with only host if no port 1`] = `"localhost/subpath"`;

exports[`BuildUrl build should return the prefix if no parameters 1`] = `"localhost:1400/subpath"`;
